hexa/datasets/models.py

```diff
@@ -0,0 +1,170 @@
+            if key in kwargs:
+                setattr(self, key, kwargs[key])
+
+        return self.save()
+
+    def delete_if_has_perm(self, *, principal: User):
+        if not principal.has_perm("datasets.delete_dataset", self):
+            raise PermissionDenied
+        self.delete()
+
+    def create_version(self, *, principal: User, name: str, description: str):
+        return DatasetVersion.objects.create_if_has_perm(
+            principal=principal,
+            dataset=self,
+            name=name,
+            description=description,
+        )
+
+    def share(self, principal: User, workspace: any):
+        if not principal.has_perm("datasets.share_dataset", self):
+            raise PermissionDenied
+
+        return DatasetShare.objects.create(
+            created_by=principal,
+            dataset=self,
+            workspace=workspace,
+        )
+
+
+class DatasetVersionQuerySet(BaseQuerySet):
+    def filter_for_user(self, user: typing.Union[AnonymousUser, User]):
+        # TODO: It should also check workspace where it's added
+        return self._filter_for_user_and_query_object(
+            user,
+            models.Q(dataset__workspace__members=user),
+            return_all_if_superuser=False,
+        )
+
+
+class DatasetVersionManager(models.Manager):
+    def create_if_has_perm(
+        self, principal: User, dataset: Dataset, *, name: str, description: str
+    ):
+        if not principal.has_perm("datasets.create_dataset_version", dataset):
+            raise PermissionDenied
+
+        version = self.create(
+            name=name,
+            dataset=dataset,
+            created_by=principal,
+            description=description,
+        )
+
+        create_bucket_folder(settings.DATASETS_BUCKET, f"{dataset.id}/{version.id}")
+
+        return version
+
+
+class DatasetVersion(Base):
+    dataset = models.ForeignKey(
+        Dataset,
+        null=False,
+        blank=False,
+        on_delete=models.CASCADE,
+        related_name="versions",
+    )
+    name = models.TextField(null=False, blank=False)
+    description = models.TextField(blank=True, null=True)
+    created_by = models.ForeignKey(User, null=True, on_delete=models.SET_NULL)
+
+    objects = DatasetVersionManager.from_queryset(DatasetVersionQuerySet)()
+
+    def delete_if_has_perm(self, *, principal: User):
+        if not principal.has_perm("datasets.delete_dataset_version", self):
+            raise PermissionDenied
+        self.delete()
+
+
+class DatasetVersionObjectQuerySet(BaseQuerySet):
+    def filter_for_user(self, user: typing.Union[AnonymousUser, User]):
+        return self._filter_for_user_and_query_object(
+            user,
+            models.Q(dataset_version__workspace__members=user)
+            | models.Q(
+                dataset__in=Dataset.objects.filter(shares__workspace__members=user)
+            ),
+            return_all_if_superuser=False,
+        )
+
+
+class DatasetVersionObject(Base):
```

Comment:
I'm still wondering about `File` vs `Object`. `Object` feels like an implementation detail.

What would you think about `File` and `uri` vs `Object` and `key`?

Reply:
Not 100% certain, so it's probably not worth changing what you implemented. In any case this can be changed later.